### PR TITLE
Allow calling protected methods from refinements

### PR DESCRIPTION
**Pull request by jhawthorn**
https://bugs.ruby-lang.org/issues/18806

Previously protected methods on refinements could never be called because they were seen as being "defined" on the hidden refinement ICLASS.

This commit updates calling refined protected methods so that they are considered to be defined on the original class (the one being refined).

This ended up using the same behaviour that was used to check whether a call to super was allowed, so I extracted that into a common method used by both.

The code inside `defined?` should probably be adjusted as well, but I will make that change separately.

### CODE REVIEW

**Review by shugo — APPROVED** [review]
Thank you!  It looks fine.

File: vm_insnhelper.c [hunk 1]
```diff
@@ -3711,6 +3711,14 @@ vm_call_method_nome(rb_execution_context_t *ec, rb_control_frame_t *cfp, struct
     }
 }
 
+static VALUE
+vm_defined_class_of_method_entry(const rb_callable_method_entry_t *me)
```
Comment (mame) [inline vm_insnhelper.c:3715]: The function name is a bit confusing. At least could you write a comment?

Comment (jhawthorn) [reply vm_insnhelper.c:3715]: 👍 It is. I will think of a clearer name.

Comment (jhawthorn) [reply vm_insnhelper.c]: Done

**Pull request merged**